### PR TITLE
[inductor] Enable FxGraphCache sharing across DDP ranks

### LINKED ISSUE #108971: Really slow compilation times for torch.compile causing distributed training errors
### 🐛 Describe the bug

We are seeing issues where `torch.compile` takes a very long time (17mins - 30 mins) to compile models despite a warm cache and results in distributed training errors like NCCL timeouts since the jobs don't make progress. A couple of examples of slow compilation are follows: https://gist.github.com/pritamdamania87/a40cbad2ec15961beb16639ddb29af9c and https://gist.github.com/pritamdamania87/c5c9031347cfb57098360ce9f93b479e.

Also, this is not the first compile where we see such slow times. `TORCHINDUCTOR_CACHE_DIR` and `TRITON_CACHE_DIR` is set and the cache was populated during the first run of the training job. The second run of job still sees slow compilation times despite also setting `torch._inductor.config.compile_threads = 16`.

Would be great to know how to debug why the compilation times are so slow and how to speed this up.

### Error logs

_No response_

### Minified repro

_No response_

### Versions

PyTorch 2.0.1

### PR DESCRIPTION
**Pull request by dhruv7539**
## Summary

Fixes https://github.com/pytorch/pytorch/issues/108971

In distributed training (DDP/FSDP), each rank uses a different CUDA device ordinal (`cuda:0` through `cuda:7`) but compiles the same model graph. The `FxGraphCache` includes device ordinals in cache keys and the generated code hardcodes device indices, preventing cache reuse across ranks. This causes every rank to independently compile through the full Dynamo→AOTAutograd→Inductor→Triton pipeline — taking 5–30 minutes on large models and triggering NCCL timeouts.

## Root Cause

Three independent issues prevent cache sharing:

1. **Cache key divergence** — `TensorMetadata.device` contains `cuda:N`, `FxGraphHashDetails.default_cuda_device_index` uses `current_device_index()`, and `CacheBase.get_system()` queries `get_device_properties(current_device())`. Each produces rank-specific hashes for identical graphs.

2. **Hardcoded device indices in generated code** — Inductor's code generator writes `torch.cuda._DeviceGuard(0)`, `torch.cuda.set_device(0)`, and `get_raw_stream(0)` into compiled Python. When rank 1 loads rank 0's cached code, tensors on `cuda:1` hit kernels targeting `cuda:0`.

3. **Device-bound kernel loading** — `_make_launchers` and `StaticCompileResult.make_launcher` load Triton `.cubin` binaries into the compile-time device's CUDA context. A CUfunction loaded on device 0 cannot be launched on device 3.

### Part 1: Cache key normalization (`codecache.py`)

Normalize CUDA device ordinals to `cuda:0` in cache key computation only:
- `extract_tensor_metadata_for_cache_key()` — normalize `TensorMetadata.device`
- `FxGraphHashDetails` — set `default_cuda_device_index = 0`
- `CacheBase.get_system()` — use `get_device_properties(0)`

### Part 2: Device-agnostic code generation (`codegen/wrapper.py`)

Replace hardcoded device indices with `torch.cuda.current_device()` in generated Python:
- `EnterDeviceContextManagerLine.codegen()` — `_DeviceGuard(current_device())`
- `write_get_raw_stream()` — `get_raw_stream(current_device())`

### Part 3: Runtime device-aware kernel loading (`runtime/triton_heuristics.py`)

Load Triton kernel binaries on the caller's actual device:
- `_make_launchers()` — use `device_interface.current_device()`
- `StaticCompileResult.make_launcher()` — use `torch.cuda.current_device()`

**Key insight:** Triton `.cubin` binaries are portable across identical GPUs — verified experimentally on 4× A100. A kernel compiled against `cuda:0` runs correctly on `cuda:3` as long as the CUfunction is loaded into the correct device's CUDA context.

## Test Results (4× A100-PCIE-40GB, PyTorch 2.10.0)

**Before fix** — cache keys diverge:
```
cuda:0 hash: 6nau7hcf7t7hhdzpq4vnuyaea4mrjiadexrma4js5gohe6erzmv
cuda:1 hash: ijdy33g47vzxnnbcyri7fisqmjw5zexhviwcel3vumeg3r56325
cuda:2 hash: 77rkpaarfvsixugrh47ff6jfymbfd7fudpvzv47dlhmz5awbims
cuda:3 hash: lb75np63mqc3rsw4j3bg4ecxlxmtnozbvk357vcvywfwihruyyn
```

**After fix** — all ranks share one cache key:
```
cuda:0 hash: 6nau7hcf7t7hhdzpq4vnuyaea4mrjiadexrma4js5gohe6erzmv
cuda:1 hash: 6nau7hcf7t7hhdzpq4vnuyaea4mrjiadexrma4js5gohe6erzmv
cuda:2 hash: 6nau7hcf7t7hhdzpq4vnuyaea4mrjiadexrma4js5gohe6erzmv
cuda:3 hash: 6nau7hcf7t7hhdzpq4vnuyaea4mrjiadexrma4js5gohe6erzmv
```

**DDP end-to-end** — all ranks compile + run correctly:
```
Rank 0: compile=3.21s  cached=0.02s  device=cuda:0
Rank 1: compile=3.28s  cached=0.02s  device=cuda:1
Rank 2: compile=3.29s  cached=0.02s  device=cuda:2
Rank 3: compile=3.20s  cached=0.02s  device=cuda:3
All outputs on correct device: True
```

## Test Plan

- [x] Verified cache key normalization: all 4 CUDA device ordinals produce identical hashes
- [x] Verified non-CUDA devices (cpu, meta) remain distinct
- [x] E2E DDP + torch.compile: all 4 ranks produce correct output on correct devices
- [x] Added `test_cuda_device_normalization_for_cache_key` unit test
- [x] Updated `test_hash_fake_tensors` to reflect normalized CUDA ordinals
- [x] Verified Triton kernels are portable across identical GPUs (4× A100)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos @ezyang @bdhirsh @anijain2305 @msaroufim @oulgen @wconstab

Made with [Cursor](https://cursor.com)

### CODE REVIEW

File: torch/_inductor/codecache.py [hunk 1]
```diff
@@ -479,6 +496,11 @@ def extract_tensor_metadata_for_cache_key(t: Tensor) -> TensorMetadata:
     if not hasattr(t, "_is_inductor_static"):
         meta = dataclasses.replace(meta, storage_offset=0, storage_bytes=None)
 
+    # Normalize CUDA device ordinals so all DDP ranks produce the same
```
Comment (wconstab) [inline torch/_inductor/codecache.py]: so, this is not really specific to DDP right? (just a nit about the comment)

but the important thing is that, regardless of DDP, FSDP or some other parallelism- there could be slight rank-to-rank divergences in the code- e.g. padding on one rank leads to a different kernel on that rank.

hopefully, we are already sufficiently caching on the src code so normalizing the device ordinal doesn't put us at risk of spurious sharing.

Comment (dhruv7539) [reply torch/_inductor/codecache.py]: Good catch on the comment — this applies to any distributed parallelism, not just DDP. Will fix.

On spurious sharing: I believe we're safe because the device ordinal is the **only** thing being normalized. All other cache key components remain intact:

- Full graph structure (via `FxGraphCachePickler` hashing the `GraphModule`)
- Tensor shapes, strides, dtypes, layouts, memory formats (all in `TensorMetadata`)
- `storage_offset`, `requires_grad`, `is_quantized`, etc.
- Kwargs, guards, deterministic algorithm settings, matmul settings
- Triton version, CUDA version, GPU model name, compute capability (via `CacheBase.get_system()`)

If one rank has different padding, its graph will have different tensor shapes/strides, producing a different hash. The device ordinal is orthogonal to the kernel's computational structure.

The only risk would be heterogeneous GPUs, but `CacheBase.get_system()` already includes GPU model name and compute capability in the system hash — so different GPU types will never share cache entries.

Comment (wconstab) [reply torch/_inductor/codecache.py]: sounds good. could you add a test showing the cache hits for the intended case, and misses when there is a slight code diff?

Comment (dhruv7539) [reply torch/_inductor/codecache.py]: Done! Added `test_cross_device_cache_sharing` in the latest commit. It tests three scenarios:

1. **Same graph, device 0** → compile + cache miss
2. **Same graph, device 1** → cache HIT (verifies cross-device sharing works)
3. **Different graph (`sin*cos` instead of `sin+cos`), device 1** → cache MISS (verifies no spurious sharing)

The test uses `@unittest.skipIf(not TEST_MULTIGPU, ...)` to ensure it only runs on multi-GPU CI nodes.

**Review by aorenste — CHANGES_REQUESTED** [review]
Lol - I was planning on getting to this eventually.

The hack we do in torchrun (`--virtual-local-rank`) is to lie to the various ranks by telling them all they're device 0 (setting LOCAL_RANK to 0) and "fix" it by munging CUDA_VISIBLE_DEVICES.

My concern with the approach in this PR is that it will work for "normal" use but if you refer to explicit devices ("cuda:3") I think it's going to break.

I think the "correct" fix is probably to route explicit DeviceId's as graph parameters instead of storing them as constants - although I'd be willing to accept other possible solutions as well.

At a minimum this needs to be gated by a config.

**Review by wconstab — COMMENTED** [review]
test LGTM.  I'll see if anyone else from the inductor team wants to give a review.  @eellison

**Comment by dhruv7539** [comment]
@aorenste Great questions — let me address both:

### (a) Cached artifact device references

The cached artifact does **not** retain `cuda:1` references. Here's what the generated Python code looks like after this PR:

```python
# Generated by our PR:
with torch.cuda._DeviceGuard(torch.cuda.current_device()):
    torch.cuda.set_device(torch.cuda.current_device())
stream0 = get_raw_stream(torch.cuda.current_device())
```

vs. what it looked like before:

```python
# Before (stock Inductor):
with torch.cuda._DeviceGuard(0):
    torch.cuda.set_device(0)
stream0 = get_raw_stream(0)
```

So the cached code is fully device-agnostic — `current_device()` is evaluated at runtime, not compile time. When rank 3 loads a cached artifact that was compiled on rank 1, it runs `current_device()` which returns 3.

Similarly, Triton kernel `.cubin` loading was changed to use `current_device()` at load time rather than the compile-time device ordinal.

**However**, you're right that I missed the `autotune_at_compile_time` code paths — lines 1529 and 1565-1572 in `wrapper.py` still embed `device_idx` directly. I'll fix those in the next push.

### (b) Multi-device graphs

You're correct — normalizing all CUDA ordinals to `cuda:0` would collapse multi-device graphs into the same cache key. This is a real limitation.

I'm happy to:
1. Document this as experimental
2. Document the single-GPU-per-rank constraint
3. Gate everything behind `config.normalize_cuda_device_for_cache` (default `False`)

For the typical distributed training case (DDP/FSDP where each rank uses one GPU), this constraint is naturally satisfied. Multi-device-per-rank graphs would simply not opt into this flag.

I'll push the config gate + autotune path fix + documentation shortly.

**Review by aorenste — APPROVED** [review]

**Review by xmfan — CHANGES_REQUESTED** [review]
The fix looks legit. But this addresses a different problem and doesn't fix the issue linked in the description, because on a fresh compile with DDP, all ranks would have no cache to share, please remove it.
